README.md
# NAME

dtgrep - print lines matching a date range

# SYNOPSIS

    dtgrep --from RFC3339 --to RFC3339 --format TIME_LAYOUT syslog

# DESCRIPTION

Do you even remember how often in your life you needed to find lines in
a log in a date range? And how often you build brittle regexps in grep
to match entries spanning over an hour change?

With dtgrep you don't have to. It features

* efficient binary search on normal files
* read bzip and gzip files without external dependencies
* automatically sort files
* merge lines from different files in output stream
* do as little work as necessary
* flexible syntax to declare date ranges

# EXAMPLES

But just let me show you a few examples.

The only parameter dtgrep really needs is _format_ to tell it how to
reckognize a timestamp. In this case dtgrep matches all lines from epoch to
the time dtgrep started.

    dtgrep --format "Jan _2 15:04:05" syslog

There are also some already predefined formats you can use:

    dtgrep --format apache access.log

You can specify which timerange to print:

    dtgrep --from 2006-01-02T12:00:00 --to 2006-01-02T12:15:00 syslog

If you leave one out it either defaults to epoch or the start of the program.

    dtgrep  --to 2006-01-02T12:15:00 --format rsyslog syslog

dtgrep can also read lines from stdin, but filtering those will be
slower as you can't just seek in a pipe.  It's often more efficient to
just redirect the lines from the pipe to a file first. But nothing is
stopping you to just call dtgrep directly.

    zcat syslog.gz | dtgrep --to 2006-01-02T12:15:00
    dtgrep --to 2006-01-02T12:15:00 syslog.gz

# OPTIONS

* --from DATESPEC

  Print all lines from RFC3339 inclusively. Defaults to January 1,
  year 1, 00:00:00 UTC. See [DATESPECS](#datespecs) for valid arguments.

* --to DATESPEC

  Print all lines until RFC3339 exclusively. Default to the current
  time. See [DATESPECS](#datespecs) for valid arguments.

* --format FORMAT

  FORMAT describes how a date looks. The first date found on a line is used.

  FORMAT can either a named format or any layout supported by the [time package](https://golang.org/pkg/time/#Parse).

  Additionally, dtgrep supports named formats:

  * rsyslog "Jan \_2 15:04:05"
  * apache "02/Jan/2006:15:04:05 -0700"
  * iso3339 "2006-01-02T15:04:05Z07:00"

  This parameter defaults to _rsyslog_.

* --multiline

  Print lines without timestamp between matching lines.

* --skip-dateless

  Ignore lines without timestamp.

* --location LOCATION

  If a date has no explicit timezone, interpret it as in the given
  location. LOCATION must be a valid location from the IANA Time Zone
  database, such as "America/New\_York".

  If the name is "" or "UTC, interpret dates as UTC.

  This parameter defaults to the system's local time zone.

* --help

  Shows a short help message

# DATESPECS

A datespec consists of a datetime and any numbers of modifiers. A
datetime can be an imcomplete date, in this case the missing values
will be filled with the current date. Without a timezone designator,
the local timezone will be used. The following formats are supported

* 04
* 15:04
* 15:04:05
* 2006-01-02 15:04
* 2006-01-02 15:04:05
* 2006-01-02 15:04:05Z07:00
* 2006-01-02T15:04:05Z07:00
* now

A modifier can either be a _truncate_ or _add_ statement. Both expect a duration as argument.

* Truncate will round the date down to the next multiple of its duration
* Add adds the duration to the current date.

Duration can be a any value parsable by the [ParseDuration function](https://golang.org/pkg/time/#ParseDuration).

Some examples

* "15:06 truncate 5m add -5m" results in 15:00 today
* "now truncate 24h add -24h" is the beginning of yesterday
* "00:00 add -24h" is also the start of the last day.
* "now"

# ENVIRONMENT

* GO\_DATEGREP\_DEFAULT\_FORMAT

  Overwrites the default for the _--format_ parameter. The syntax is described there.

# LIMITATION

dtgrep expects the files to be sorted. If the timestamps are not
ascending, dtgrep might be exiting before the last line in its date
range is printed. So, no daylight saving time.

# SEE ALSO

This is by far no new idea. Just on github you can find at least 10
programs that at least the binary search part of the problem. Search
for tgrep, timegrep or dategrep.

I wrote [dategrep](http://github.com/mdom/dategrep) without knowing
about these. What is weird, as  I distinctly remember that i searched
extensively for such a program but I must have choosen very poor terms.

Since then i never lost interest in the problem. This is another try to
find the right mix of features, easy deployment and faste results.

# COPYRIGHT AND LICENSE

Copyright 2016 Mario Domgoergen <mario@domgoergen.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

### Added

- Additional formats for --from and --to

### Fixed

- The time used for "now" is set once and evaluates always to the same time.

### Changed
### Deprecated
### Removed
### Security

## [1.1.0] - 2016-09-05

### Added

- Add --version switch
- Allow date range with offsets and truncation
- Add --duration

### Fixed

- Check if start date is before end date

## [1.0.0] - 2016-08-22

### Added

- Use efficient binary search on normal files
- Read bzip and gzip files without external dependencies
- Sort files automatically
- Merge lines from different files in output stream



main_test.go
package main

import (
	"github.com/mdom/dtgrep/fixtime"
	"testing"
	"time"
)

type FillDateTest struct {
	argument string
	result   string
}

func TestAddYear(t *testing.T) {

	now, _ := time.Parse(time.RFC3339, "2006-01-02T15:04:05Z")

	tests := []FillDateTest{
		{"0000-01-02T15:04:05Z", "2006-01-02T15:04:05Z"},
		{"0000-01-03T15:04:05Z", "2005-01-03T15:04:05Z"},
	}

	for _, v := range tests {
		argument, _ := time.Parse(time.RFC3339, v.argument)
		result, _ := time.Parse(time.RFC3339, v.result)
		if fixtime.AddYear(argument, now) != result {
			t.Error("fillDate failed", argument, result)
		}
	}

}

func TestDateRange(t *testing.T) {
	from, _ := time.Parse(time.RFC3339, "2016-05-09T10:40:00Z")
	to, _ := time.Parse(time.RFC3339, "2016-05-09T11:40:00Z")

	var s, e time.Time
	var d time.Duration

	s, e = dateRange(from, to, time.Duration(0))
	if s.String() != "2016-05-09 10:40:00 +0000 UTC" || e.String() != "2016-05-09 11:40:00 +0000 UTC" {
		t.Error("specified from and to without duration")
	}

	d, _ = time.ParseDuration("20s")
	s, e = dateRange(from, time.Time{}, d)
	if s.String() != "2016-05-09 10:40:00 +0000 UTC" || e.String() != "2016-05-09 10:40:20 +0000 UTC" {
		t.Error("specified from with duration")
	}

	d, _ = time.ParseDuration("20s")
	s, e = dateRange(time.Time{}, to, d)
	if s.String() != "2016-05-09 11:39:40 +0000 UTC" || e.String() != "2016-05-09 11:40:00 +0000 UTC" {
		t.Error("specified to with duration")
	}

}


retime/retime_test.go
package retime

import (
	"testing"
	"time"
)

func TestCompileToRegexp(t *testing.T) {
	f, _ := New("Jan _2 15:04:05", time.UTC)

	dt1, _ := time.ParseInLocation("Jan _2 15:04:05", "Feb 12 09:34:59", time.UTC)
	dt2, _ := f.Extract("foo Feb 12 09:34:59 bar")

	if !dt2.Equal(dt1) {
		t.Error("foo")
	}
}


dateflag/dateflag_test.go
package dateflag

import (
	"testing"
	"time"
)

func TestDateFlagSet(t *testing.T) {
	time.Local = time.UTC
	now, _ := time.Parse(time.RFC3339, "2016-05-09T10:40:00Z")
	d := &DateFlag{Now: now}

	var err error

	err = d.Set("12")
	if err != nil || d.String() != "2016-05-09 10:12:00 +0000 UTC" {
		t.Error("Passing 12 failed")
	}

	err = d.Set("12:15")
	if err != nil || d.String() != "2016-05-09 12:15:00 +0000 UTC" {
		t.Error("Passing 12:15 failed")
	}

	err = d.Set("2016-05-08 12:15")
	if err != nil || d.String() != "2016-05-08 12:15:00 +0000 UTC" {
		t.Error("Passing 2016-05-08 12:15 failed")
	}

	err = d.Set("12:15 truncate 1h")
	if err != nil || d.String() != "2016-05-09 12:00:00 +0000 UTC" {
		t.Error("Passing 12:15 truncate 1h failed")
	}

	err = d.Set("12:15 add 5m")
	if err != nil || d.String() != "2016-05-09 12:20:00 +0000 UTC" {
		t.Error("Passing 12:15 add 5m failed")
	}

	err = d.Set("now")
	if err != nil || d.String() != "2016-05-09 10:40:00 +0000 UTC" {
		t.Error("Passing now failed")
	}

	err = d.Set("truncate 1h")
	if err != nil || d.String() != "2016-05-09 10:00:00 +0000 UTC" {
		t.Error("Passing truncate 1h failed")
	}

	err = d.Set("12:15 minus 5m")
	if err == nil || d.String() == "2016-05-09 12:10:00 +0000 UTC" {
		t.Error("Passing unknown operator minus succeeded")
	}

	err = d.Set("12:15 truncate 1h minus 5m")
	if err == nil || d.String() == "2016-05-09 11:55:00 +0000 UTC" {
		t.Error("Passing 12:15 truncate 1h minus 5m succeeded")
	}

	err = d.Set("12:15 truncate")
	if err == nil {
		t.Error("Passing 12:15 truncate without argument succeeded")
	}

	err = d.Set("12:15 truncate hour")
	if err == nil {
		t.Error("Passing 12:15 truncate hour succeeded")
	}

}


format/format.go
package format

import (
	"regexp"
	"bytes"
)

type Format struct {
	regexp   string
	name     string
	template string
}

func CompileFormatToRegexp(layout string) string {
	var buffer bytes.Buffer

	l := len(layout)
	for i := 0; i < l; {
		switch {
		case l >= i+7 && layout[i:i+7] == "January":
			buffer.WriteString(`[A-Z][a-z]{3,9}`)
			i += 7
		case l >= i+3 && layout[i:i+3] == "Jan":
			buffer.WriteString(`[A-Z][A-Za-z]{2}`)
			i += 3
		case l >= i+6 && layout[i:i+6] == "Monday":
			buffer.WriteString(`[A-Z][a-z]{6,9}`)
			i += 6
		case l >= i+3 && layout[i:i+3] == "Mon":
			buffer.WriteString(`[A-Z][A-Za-z]{2}`)
			i += 3
		case l >= i+3 && layout[i:i+3] == "MST":
			buffer.WriteString(`[A-Z]+`)
			i += 3
		// 01, 02, 03, 04, 05, 06
		case l >= i+1 && layout[i] == '0' && '1' <= layout[i+1] && layout[i+1] <= '6':
			buffer.WriteString(layout[i:i+1])
			i += 2
		case l >= i+1 && layout[i:i+1] == "15": //hour
			buffer.WriteString(`\d\d`)
			i += 2
		case layout[i] == '1': // month
			buffer.WriteString(`\d?\d`)
			i++
		case l >= i+4 && layout[i:i+4] == "2006": //year
			buffer.WriteString(`\d{4}`)
			i += 4
		case layout[i] == '2': // day
			buffer.WriteString(`\d`)
			i++
		case l >= i+5 && layout[i:i+5] == "_2006":
			buffer.WriteString(`_\d{4}`)
			i += 5
		case l >= i+2 && layout[i:i+1] == "_2": // day
			buffer.WriteString(`[ \d]\d`)
			i += 2
		case  layout[i] == '3': // hour12
			buffer.WriteString(`\d?\d`)
			i++
		case layout[i] == '4': // minute
			buffer.WriteString(`\d?\d`)
			i++
		case layout[i] == '5': // second
			buffer.WriteString(`\d?\d`)
			i++
		case l > i+1 && layout[i:i+1] == "PM":
			buffer.WriteString(`(PM|AM)`)
			i += 2
		case l > i+1 && layout[i:i+1] == "pm":
			buffer.WriteString(`(pm|am)`)
			i += 2
		case l >= i+7 && layout[i:i+7] == "-070000":
			buffer.WriteString(`[+-]\d{7}`)
			i += 7
		case l >= i+9 && layout[i:i+9] == "-07:00:00":
			buffer.WriteString(`[+-]\d\d:\d\d:\d\d`)
			i += 9
		case l >= i+5 && layout[i:i+5] == "-0700":
			buffer.WriteString(`[+-]\d{4}`)
			i += 5
		case l >= i+6 && layout[i:i+6] == "-07:00" :
			buffer.WriteString(`[+-]\d\d:\d\d`)
			i += 6
		case l >= i+3 && layout[i:i+3] == "-07" :
			buffer.WriteString(`[+-]\d{2}`)
			i += 3
		case l >= i+7 && layout[i:i+7] == "Z070000":
			buffer.WriteString(`(Z|[+-]\d{7})`)
			i += 7
		case l >= i+9 && layout[i:i+9] == "Z07:00:00":
			buffer.WriteString(`(Z|[+-]\d\d:\d\d:\d\d)`)
			i += 9
		case l >= i+5 && layout[i:i+5] == "Z0700":
			buffer.WriteString(`(Z|[+-]\d{4})`)
			i += 5
		case l >= i+6 && layout[i:i+6] == "Z07:00":
			buffer.WriteString(`(Z|[+-]\d\d:\d\d)`)
			i += 6
		case l >= i+3 && layout[i:i+3] == "Z07":
			buffer.WriteString(`(Z|[+-]\d{2})`)
			i += 3
		case l >= i+1 && ( layout[i:i+2] == ".0" || layout[i:i+2] == ".9"):
			j := i + 1
			ch := layout[i+2]
			for j < l && layout[j] == ch {
				j++
			}
			buffer.WriteString(`\.\d+`)
			i = i + j +1
		default:
			buffer.WriteString(regexp.QuoteMeta(string(layout[i])))
			i++
		}
	}
	return buffer.String()
}


retime/retime.go
package retime

import (
	"bytes"
	"regexp"
	"time"
)

type Format struct {
	regexp *regexp.Regexp
	layout string
	loc    *time.Location
}

func New(layout string, loc *time.Location) (Format, error) {
	format := Format{
		layout: layout,
		loc:    loc,
	}
	regexp, err := compileToRegexp(layout)
	if err != nil {
		return format, err
	}
	format.regexp = regexp
	return format, nil
}

func (f *Format) Extract(s string) (time.Time, error) {
	match := f.regexp.FindString(s)
	return time.ParseInLocation(f.layout, match, f.loc)
}

func prefixAt(s string, index int, prefix string) bool {
	return len(s) >= index+len(prefix) && s[index:index+len(prefix)] == prefix
}

func compileToRegexp(layout string) (*regexp.Regexp, error) {
	var buffer bytes.Buffer

	l := len(layout)
	for i := 0; i < l; {
		switch {
		case prefixAt(layout, i, "January"):
			buffer.WriteString(`[A-Z][a-z]{3,9}`)
			i += 7
		case prefixAt(layout, i, "Jan"):
			buffer.WriteString(`[A-Z][A-Za-z]{2}`)
			i += 3
		case prefixAt(layout, i, "Monday"):
			buffer.WriteString(`[A-Z][a-z]{6,9}`)
			i += 6
		case prefixAt(layout, i, "Mon"):
			buffer.WriteString(`[A-Z][A-Za-z]{2}`)
			i += 3
		case prefixAt(layout, i, "MST"):
			buffer.WriteString(`[A-Z]+`)
			i += 3
		// 01, 02, 03, 04, 05, 06
		case l >= i+1 && layout[i] == '0' && '1' <= layout[i+1] && layout[i+1] <= '6':
			buffer.WriteString(`\d\d`)
			i += 2
		case prefixAt(layout, i, "15"):
			buffer.WriteString(`\d\d`)
			i += 2
		case layout[i] == '1': // month
			buffer.WriteString(`\d?\d`)
			i++
		case prefixAt(layout, i, "2006"):
			buffer.WriteString(`\d{4}`)
			i += 4
		case layout[i] == '2': // day
			buffer.WriteString(`\d`)
			i++
		case prefixAt(layout, i, "15"):
			buffer.WriteString(`_\d{4}`)
			i += 5
		case prefixAt(layout, i, "_2"): //day
			buffer.WriteString(`[ \d]\d`)
			i += 2
		case layout[i] == '3': // hour12
			buffer.WriteString(`\d?d`)
			i++
		case layout[i] == '4': // minute
			buffer.WriteString(`\d\d`)
			i++
		case layout[i] == '5': // second
			buffer.WriteString(`\d\d`)
			i++
		case prefixAt(layout, i, "PM"):
			buffer.WriteString(`(PM|AM)`)
			i += 2
		case prefixAt(layout, i, "pm"):
			buffer.WriteString(`(pm|am)`)
			i += 2
		case prefixAt(layout, i, "-070000"):
			buffer.WriteString(`[+-]\d{7}`)
			i += 7
		case prefixAt(layout, i, "-07:00:00"):
			buffer.WriteString(`[+-]\d\d:\d\d:\d\d`)
			i += 9
		case prefixAt(layout, i, "-0700"):
			buffer.WriteString(`[+-]\d{4}`)
			i += 5
		case prefixAt(layout, i, "-07:00"):
			buffer.WriteString(`[+-]\d\d:\d\d`)
			i += 6
		case prefixAt(layout, i, "-07"):
			buffer.WriteString(`[+-]\d{2}`)
			i += 3
		case prefixAt(layout, i, "Z070000"):
			buffer.WriteString(`(Z|[+-]\d{7})`)
			i += 7
		case prefixAt(layout, i, "Z07:00:00"):
			buffer.WriteString(`(Z|[+-]\d\d:\d\d:\d\d)`)
			i += 9
		case prefixAt(layout, i, "Z0700"):
			buffer.WriteString(`(Z|[+-]\d{4})`)
			i += 5
		case prefixAt(layout, i, "Z07:00"):
			buffer.WriteString(`(Z|[+-]\d\d:\d\d)`)
			i += 6
		case prefixAt(layout, i, "Z07"):
			buffer.WriteString(`(Z|[+-]\d{2})`)
			i += 3
		case l >= i+1 && (layout[i:i+1] == ".0" || layout[i:i+1] == ".9"):
			j := i + 1
			ch := layout[i+2]
			for j < l && layout[j] == ch {
				j++
			}
			buffer.WriteString(`\.\d+`)
			i = i + j + 1
		default:
			buffer.WriteString(regexp.QuoteMeta(string(layout[i])))
			i++
		}
	}
	return regexp.Compile(buffer.String())
}


dateflag/dateflag.go
package dateflag

import (
	"errors"
	"github.com/mdom/dtgrep/fixtime"
	"regexp"
	"strings"
	"time"
)

type formats struct {
	template string
	complete func(date time.Time, now time.Time) time.Time
}

type DateFlag struct {
	date, Now time.Time
}

func (d *DateFlag) String() string {
	return d.date.String()
}

func (d *DateFlag) Get() time.Time {
	return d.date
}

func (d *DateFlag) Set(dateSpec string) error {

	datePart := dateSpec
	modPart := ""

	if d.Now.IsZero() {
		d.Now = time.Now()
	}

	results := regexp.MustCompile(`add|truncate`).FindStringIndex(dateSpec)
	if results != nil {
		idx := results[0]
		if idx == 0 {
			datePart = ""
			modPart = dateSpec
		} else {
			datePart = dateSpec[:idx-1]
			modPart = dateSpec[idx:]
		}
	}

	var modifiers []func(time.Time) time.Time
	fields := strings.Fields(modPart)

	for i := 0; i < len(fields); i += 2 {

		if i+1 == len(fields) {
			return errors.New("Missing argument for " + fields[i])
		}

		d, err := time.ParseDuration(fields[i+1])
		if err != nil {
			return err
		}
		switch fields[i] {
		case "truncate":
			modifiers = append(modifiers, func(t time.Time) time.Time { return t.Truncate(d) })
		case "add":
			modifiers = append(modifiers, func(t time.Time) time.Time { return t.Add(d) })
		default:
			return errors.New("Unknown operator " + fields[i])
		}
	}

	var dt time.Time

	if datePart == "now" || datePart == "" {
		dt = d.Now
	} else {
		specs := []formats{
			{"04", fixtime.AddDateHour},
			{"15:04", fixtime.AddDate},
			{"15:04:05", fixtime.AddDate},
			{"2006-01-02 15:04", nil},
			{"2006-01-02 15:04:05", nil},
			{"2006-01-02 15:04:05Z07:00", nil},
			{time.RFC3339, nil},
		}
		var err error
		for _, spec := range specs {
			dt, err = time.ParseInLocation(spec.template, datePart, time.Local)
			if err == nil {
				if spec.complete != nil {
					dt = spec.complete(dt, d.Now)
				}
				break
			}
		}
		if err != nil {
			return err
		}
	}
	for _, mod := range modifiers {
		dt = mod(dt)
	}
	d.date = dt
	return nil
}


main.go
package main

import (
	"bufio"
	"compress/bzip2"
	"compress/gzip"
	"flag"
	"fmt"
	"github.com/mdom/dtgrep/dateflag"
	"github.com/mdom/dtgrep/fixtime"
	"github.com/mdom/dtgrep/retime"
	"io"
	"log"
	"os"
	"path"
	"sort"
	"time"
)

var now = time.Now()
var epoch time.Time
var loc = time.Local

var Version = "unknown"
var CommitHash = "unknown"
var BuildDate = "unknown"

type Options struct {
	from, to     time.Time
	skipDateless bool
	multiline    bool
}

type Iterator struct {
	filename string
	reader   io.Reader
	*bufio.Scanner
	Line string
	Time time.Time
	Err  error
}

type Iterators []*Iterator

func (it Iterators) Len() int           { return len(it) }
func (it Iterators) Swap(i, j int)      { it[i], it[j] = it[j], it[i] }
func (it Iterators) Less(i, j int) bool { return it[i].Time.Before(it[j].Time) }

func inTimeRange(s *Iterator, from, to time.Time) bool {
	dt := s.Time
	return (dt.Equal(from) || dt.After(from)) && dt.Before(to)
}

func filter(s Iterators, from, to time.Time) Iterators {
	var p Iterators
	for _, v := range s {
		if v.Err == nil && inTimeRange(v, from, to) {
			p = append(p, v)
		}
	}
	return p
}

var formats = map[string]string{
	"rsyslog": "Jan _2 15:04:05",
	"rfc3339": time.RFC3339,
	"apache":  "02/Jan/2006:15:04:05 -0700",
}

func dateRange(from, to time.Time, duration time.Duration) (time.Time, time.Time) {

	// --duration, --from and --to specified
	if duration != 0 && !to.IsZero() && !from.IsZero() {
		log.Fatalln("--duration can only be used with either --from or --to.")
	}

	// only --duration specified
	if duration != 0 && to.IsZero() && from.IsZero() {
		switch {
		case duration.Hours() >= 1:
			to = now.Truncate(time.Duration(1) * time.Hour)
		case duration.Minutes() >= 1:
			to = now.Truncate(time.Duration(1) * time.Minute)
		default:
			to = now.Truncate(time.Duration(1) * time.Second)
		}
		from = to.Add(-duration)
	}

	if duration != 0 && !to.IsZero() && from.IsZero() {
		from = to.Add(-duration)
	}

	if duration != 0 && to.IsZero() && !from.IsZero() {
		to = from.Add(duration)
	}

	if to.IsZero() {
		to = now
	}

	return from, to
}

func main() {

	log.SetFlags(0)
	log.SetPrefix("")

	var formatName, location string

	toFlag := dateflag.DateFlag{Now: now}
	fromFlag := dateflag.DateFlag{Now: now}

	var duration time.Duration

	var options Options

	defaultFormat := "rsyslog"
	if os.Getenv("GO_DATEGREP_FORMAT") != "" {
		defaultFormat = os.Getenv("GO_DATEGREP_FORMAT")
	}

	flag.Var(&fromFlag, "from", "Print all lines from `DATESPEC` inclusively.")
	flag.Var(&toFlag, "to", "Print all lines until `DATESPEC` exclusively.")

	flag.StringVar(&formatName, "format", defaultFormat, "Use `FORMAT` to parse file.")
	flag.BoolVar(&options.skipDateless, "skip-dateless", false, "Ignore all lines without timestamp.")
	flag.BoolVar(&options.multiline, "multiline", false, "Print all lines between the start and end line even if they are not timestamped.")
	flag.StringVar(&location, "location", time.Local.String(), "Use location in the absence of any timezone information.")

	flag.DurationVar(&duration, "duration", 0, "Print all lines in `DURATION` from --from or --to.")

	var displayVersion bool
	flag.BoolVar(&displayVersion, "version", false, "Display version")

	flag.Lookup("to").DefValue = "now"
	flag.Lookup("from").DefValue = "epoch"

	flag.Parse()

	if displayVersion {
		log.Printf("version: %s\ncommit: %s\nbuild date: %s\n",
			Version, CommitHash, BuildDate)
		return
	}

	var err error

	loc, err = time.LoadLocation(location)
	if err != nil {
		log.Fatalln("Can't load location:", err)
	}

	options.from, options.to = dateRange(fromFlag.Get(), toFlag.Get(), duration)

	if options.from.After(options.to) || options.from.Equal(options.to) {
		log.Fatalln("Start date must be before end date.")
	}

	var format retime.Format
	for name, template := range formats {
		if name == formatName {
			format, err = retime.New(template, loc)
			if err != nil {
				log.Fatalln("Can't create format:", err)
			}
			break
		}
	}

	if (format == retime.Format{}) {
		format, err = retime.New(formatName, loc)
		if err != nil {
			log.Fatalln("Can't create format:", err)
		}
	}

	var iterators = make(Iterators, 0)

	if len(flag.Args()) > 0 {
		for _, filename := range flag.Args() {

			if filename == "-" {
				i := &Iterator{filename: filename, reader: os.Stdin, Scanner: bufio.NewScanner(os.Stdin)}
				iterators = append(iterators, i)
				continue
			}

			file, err := os.Open(filename)
			if err != nil {
				log.Fatalln("Cannot open", filename, ":", err)
			}
			defer file.Close()

			// mimeType support?
			ext := path.Ext(filename)
			if ext == ".gz" || ext == ".z" {
				r, err := gzip.NewReader(file)
				defer r.Close()
				if err != nil {
					log.Fatalln("Cannot open", filename, ":", err)
				}
				i := &Iterator{filename: filename, reader: r, Scanner: bufio.NewScanner(r)}
				iterators = append(iterators, i)
			} else if ext == ".bz2" || ext == ".bz" {
				r := bzip2.NewReader(file)
				i := &Iterator{filename: filename, reader: r, Scanner: bufio.NewScanner(r)}
				iterators = append(iterators, i)
			} else {
				scanner, err := findStartSeekable(file, options, format)
				switch {
				case err == io.EOF:
					// daterange not in file, skip
					continue
				case err != nil:
					log.Fatalln("Error finding dates in ", filename, ":", err)
				}
				i := &Iterator{filename: filename, reader: file, Scanner: scanner}
				iterators = append(iterators, i)
			}
		}
	} else {
		i := &Iterator{filename: "-", reader: os.Stdin, Scanner: bufio.NewScanner(os.Stdin)}
		iterators = append(iterators, i)
	}

	var ignoreError = options.skipDateless || options.multiline
	for _, i := range iterators {
		i.Scan(options.from, options.to, ignoreError, format)
	}

	for {

		iterators = filter(iterators, options.from, options.to)
		sort.Sort(iterators)

		if len(iterators) > 0 {
			var until time.Time
			if len(iterators) > 1 {
				until = iterators[1].Time
			} else {
				until = options.to
			}
			i := iterators[0]
			fmt.Println(i.Line)
			i.Print(until, options, format)
		} else {
			break
		}
	}
}

func (i *Iterator) Print(to time.Time, options Options, format retime.Format) {
	for {
		i.Line, i.Err = readline(i.Scanner)
		if i.Err == io.EOF {
			return
		}
		if i.Err != nil {
			// what file?
			log.Fatalln("Error reading file:", i.Err)
		}
		i.Time, i.Err = format.Extract(i.Line)
		i.Time = fixtime.AddYear(i.Time, now)

		switch {
		case i.Err != nil && options.multiline:
			fmt.Println(i.Line)
		case i.Err != nil && options.skipDateless:
			continue
		case i.Err != nil:
			log.Fatalln("Aborting. Found line without date:", i.Line)
		case i.Time.Before(to):
			fmt.Println(i.Line)
		default:
			return
		}
	}
}

func readline(s *bufio.Scanner) (string, error) {
	ret := s.Scan()
	if !ret && s.Err() == nil {
		return "", io.EOF
	}
	if !ret {
		return "", s.Err()
	}
	return s.Text(), nil
}

func (i *Iterator) Scan(from, to time.Time, ignoreError bool, format retime.Format) {
	for {
		i.Line, i.Err = readline(i.Scanner)
		if i.Err != nil {
			break
		}
		i.Time, i.Err = format.Extract(i.Line)
		i.Time = fixtime.AddYear(i.Time, now)
		if i.Err != nil && ignoreError {
			continue
		}
		if i.Err != nil {
			log.Fatalln("Aborting. Found line without date:", i.Line)
		}
		if i.Time.After(to) {
			i.Err = io.EOF
			break
		}
		if i.Time.Equal(from) || i.Time.After(from) {
			break
		}
	}
}

func findStartSeekable(f *os.File, options Options, format retime.Format) (*bufio.Scanner, error) {

	// find block size
	blockSize := int64(4096)

	fileInfo, err := f.Stat()
	if err != nil {
		return &bufio.Scanner{}, err
	}
	size := fileInfo.Size()
	min := int64(0)
	max := size / blockSize
	var mid int64

	var ignoreErrors = options.skipDateless || options.multiline

	for max-min > 1 {
		mid = (max + min) / 2
		f.Seek(mid*blockSize, os.SEEK_SET)
		scanner := bufio.NewScanner(f)

		_, err := readline(scanner) // skip partial line
		if err != nil {
			return scanner, err
		}

		var dt time.Time

		for {
			line, err := readline(scanner)
			if err != nil {
				return scanner, err
			}

			dt, err = format.Extract(line)
			dt = fixtime.AddYear(dt, now)
			if err != nil && ignoreErrors {
				continue
			}
			if err != nil {
				log.Fatalln("Aborting. Found line without date:", line)
			}
			break
			// optimization: while searching next line we entered next block
		}

		if dt.Before(options.from) {
			min = mid
		} else {
			max = mid
		}
	}

	min = min * blockSize
	_, err = f.Seek(min, os.SEEK_SET)
	if err != nil {
		return &bufio.Scanner{}, err
	}
	scanner := bufio.NewScanner(f)
	if min > 0 {
		_, err := readline(scanner) // skip partial line
		if err != nil {
			return scanner, err
		}
	}

	return scanner, nil
}
